wallet: Ensure best block matches wallet scan state

achow101 opened the pull request:
Implements the idea discussed in https://github.com/bitcoin/bitcoin/pull/29652#issuecomment-2010579484

Currently, `m_last_block_processed` and `m_last_block_processed_height` are not guaranteed to match the block locator stored in the wallet, nor do either of those fields actually represent the last block that the wallet is synced up to. This is confusing and unintuitive.

This PR changes those last block fields to be updated whenever the wallet makes a change to the db for new transaction state found in new blocks. Whenever a block is received that contains a transaction relevant to the wallet, the last block locator will now be written to disk. Furthermore, every block disconnection will now write an updated locator.

To ensure that the locator is relatively recent and loading rescans are fairly quick in the event of unplanned shutdown, it is also now written every 144 blocks (~1 day). Additionally it is now written when the wallet is unloaded so that it is accurate when the wallet is loaded again.

Lastly, the `chainstateFlushed` notification in the wallet is changed to be a no-op. The best block locator record is no longer written when `chainstateFlushed` is received from the node since it should already be mostly up to date.

achow101 committed:
wallet: Update best block record after block dis/connect

When a block is connected, if the new block had anything relevant to the
wallet, update the best block record on disk. If not, also sync the best
block record to disk every 144 blocks.

Also reuse the new WriteBestBlock method in BackupWallet.

achow101 committed:
wallet: Replace chainStateFlushed in loading with SetLastBlockProcessed

The only reason to call chainStateFlushed during wallet loading is to
ensure that the best block is written. Do these writes explicitly to
prepare for removing chainStateFlushed, while also ensuring that the
wallet's in memory state tracking is written to disk.

Additionally, after rescanning on wallet loading, instead of writing
the locator for the current chain tip, write the locator for the last
block that the rescan had scanned. This ensures that the stored best
block record matches the wallet's current state.

Any blocks dis/connected during the rescan are processed after the
rescan and the last block processed will be updated accordingly.

achow101 committed:
wallet: Remove chainStateFlushed

chainStateFlushed is no longer needed since the best block is updated
after a block is scanned. Since the chainstate being flushed does not
necessarily coincide with the wallet having processed said block, it
does not entirely make sense for the wallet to be recording that block
as its best block, and this can cause race conditions where some blocks
are not processed. Thus, remove this notification.

ryanofsky commented:
I'm not sure, but it seems like a potentially good thing to me that `last_block_processed` and BESTBLOCK are two distinct concepts.

The last block processed is the block that CWallet in-memory state has been synced to (particularly CWalletTx::m_state which includes mempool / abandoned information not serialized to disk).

The BESTBLOCK record is the last block whose data has been flushed to disk, that the wallet should begin syncing from next time it is reloaded.

It seems like the first commit ae3b8280a7dbb6cc87814ed97dbea5a122cf8215 could lead to a performance regression if it is writing the flushing the BESTBLOCK record on every single blockConnected event from the node. For example during reindexing it could write and sync the database each time a block is connected, even if the block is before the wallets birthday, or doesn't have any relevant transactions?

I think it probably makes sense to write BESTBLOCK in a smarter way, and probably write it more frequently, but it seems unnecessary to have to write it every block, and it might be bad for performance now or make optimizations harder in the future.

I'm also curious about the idea of saving height with the best block record. I could imagine this being useful, since IIRC parts of wallet loading are complicated by not knowing heights of transactions before the chain is attached, but it doesn't really seem like the height is used for much yet. Was this intended to be used by something else later?

achow101 commented:
> The last block processed is the block that CWallet in-memory state has been synced to (particularly CWalletTx::m_state which includes mempool / abandoned information not serialized to disk).

But none of that state is relevant to the last block processed. Any state related to blocks (confirmed or conflicted) is written to disk.

> The BESTBLOCK record is the last block whose data has been flushed to disk, that the wallet should begin syncing from next time it is reloaded.

The point is that this discrepancy can result in skipping blocks. It doesn't make sense to me that we wouldn't store the block the wallet's state has been synced to, and it doesn't make sense that we should rely on a field that means something else to determine what point to start rescanning from on the next load.

If BESTBLOCK doesn't match the chainstate, that's fine because it's a locator and we will just find the fork and sync from there. I don't think it's necessary to record which block we think the chainstate is synced to.

ryanofsky commented:
> It doesn't make sense to me that we wouldn't store the block the wallet's state has been synced to

You may be right this is the better approach. But I think the previous approach does make some sense, too. You might not want to write to the wallet database each time a block is connected if the block doesn't contain any relevant transactions, especially during reindexing.

achow101 commented:
> You might not want to write to the wallet database each time a block is connected if the block doesn't contain any relevant transactions

I think you would since not writing it would result in possibly rescanning that block at the next loading which takes a little bit of time, regardless of whether any relevant transactions are in the block.

> especially during reindexing.

Perhaps an easy solution to this is to just write the best block every 1000 blocks (or some other interval) when we are in IBD?

> I'm also curious about the idea of saving height with the best block record. I could imagine this being useful, since IIRC parts of wallet loading are complicated by not knowing heights of transactions before the chain is attached, but it doesn't really seem like the height is used for much yet. Was this intended to be used by something else later?

This was mainly done to avoid looking up the height every time we load since that was being problematic and causing a tests to fail. But it definitely could be more useful in the future.

ryanofsky commented:
The idea of saving heights is interesting to me because wallet code assumes it know block heights many of places, but it doesn't actually store heights anywhere. So for example, if there wallet stored a mapping of block hashes to heights (or other block information) it might be useful for allowing the wallet to work in an offline mode or letting the wallet CLI tool have more capabilities. This is all pretty abstract though, so I'm not suggesting something specific.

> Perhaps an easy solution to this is to just write the best block every 1000 blocks (or some other interval) when we are in IBD?

Yes but I think that just adds back the complexity you were trying to get rid of, in a form that seems worse than the status quo. If you implement that, the wallet will be back to tracking the last block processed separately from the best block written to the database. And now, instead of the node determining when data should be flushed to disk and having all wallets flush that data simultaneously, each wallet will have internal logic deciding when to flush. This would be more complicated, and could be worse for power consumption and performance if there are multiple wallets and flushes are happening more frequently at different times.

I think overall this PR is doing some good things, but the goals seem either not clear, or not clearly good, so I wonder if maybe you could take these changes and make more targeted PRs for the things you want to achieve?

Or, if this PR is definitely the direction you want to go, I'm happy to review it. I don't like some things it is doing, but overall I think it is a reasonable change.

fjahr reviewed (commented):
Concept ACK

I can confirm that best block locator and `last_processed_block` being different is confusing, see also https://github.com/bitcoin/bitcoin/pull/30455#discussion_r1719951882

Currently, this needs a rebase and I'm curious if @achow101 plans to make further changes based on @ryanofsky 's comments.

ryanofsky commented:
> I can confirm that best block locator and last_processed_block being different is confusing

I wonder if something else could be done to resolve this confusion other than writing to every wallet file every time a new block is connected, even if the block doesn't contain any relevant transactions. To me, "last block processed" and "last block flushed" seem like different concepts, and we could force them to be the same but only if we:

- Give up flexibility of not needing to flush each wallet each time a block is connected (which seems inefficient during sync)
- Give up ability to do coordinated flushes across the chainstate database, index databases and wallet databases time (which seems like it could waste resources and hurt system performance)

fjahr commented:
> I wonder if something else could be done to resolve this confusion

Ok, I would say it's a problem when users see behavior that is inconsistent because of this difference. If the users don't notice it I guess these don't have to be the same at all times. There could be a more "pull-based" approach where they are aligned when the user interacts with the wallet, similar to what I tried to do for the backup case in https://github.com/bitcoin/bitcoin/pull/30678. But I am not clear yet on which approach is the better one all factors considered.

ryanofsky commented:
It would be great if we could open an issue describing the user behavior that's confusing. I read the comment you linked to but I didn't really understand the context of what was going on with those backups. If there can be a github issue that simply describes steps to reproduce, expected behavior, and actual behavior, we can probably figure out if expected behavior is reasonable, and implement a fix if so.

fjahr commented:
> It would be great if we could open an issue describing the user behavior that's confusing.

I have created an issue here: #30686

Thompson1985 reviewed (approved):
Thank you 👍

mzumsande reviewed (commented):
Concept ACK

I verified that this would solve the issues from #31824, only looked at the first couple commits in detail so far.

With respect to the above discussion, I think that the concept of delayed flushing only makes sense if it applies to all relevant data for a given block connection /disconnection (which is the case for the indexes, for example). But when other changes to wallet txns are synced to disk immediately, it seems wrong to delay flushing the locator - this will lead to a mixed state where some relevant data is flushed and other data isn't, which will lead to issues such as the one above.

As for the point of IBD performance due to updating the locator at every block: Maybe this could be prevented by keeping track during block (dis)connection, if we did any other changes to the wallet db for this block. If nothing else needed changing ( wallet not affected by the block, only the locator would be updated) it should be safe to delay / skip the update.

achow101 commented:
As discussed at CoreDev, I've updated this PR to sync the best block periodically or when a connected block contains a relevant transaction.

I've also had to make a couple of changes to `AttachChain` following #31629

achow101 committed:
wallet: Write best block record on unload

achow101 committed:
wallet: Remove unnecessary database Close step on shutdown

StopWallets, which was being called prior to UnloadWallets, performs an
unnecessary database closing step. This causes issues in UnloadWallets
which does additional database cleanups. Since the database closing step
is unnecessary, StopWallets is removed, and UnloadWallets is now called
from WalletLoaderImpl::stop.

achow101 commented:
I added a change to make sure that the most recent best block is written to disk on wallet unload. However, it seems like this uncovered some weirdness in the wallet unloading on shutdown. In particular, `StopWallets()` was being called before `UnloadWallets()` which I think is incorrect. I added a commit which flips this order, and it seems to work, although there may be side effects I have not discovered yet. This primarily affects BDB wallets as the shutdown for BDB wallets is much more involved and confusing due to the shared database environment. This is not an issue for SQLite.

mzumsande commented on src/wallet/wallet.cpp:
```diff
@@ -649,15 +649,6 @@ bool CWallet::ChangeWalletPassphrase(const SecureString& strOldWalletPassphrase,
     return false;
 }
 
-void CWallet::chainStateFlushed(ChainstateRole role, const CBlockLocator& loc)
```
Two calls to `chainStateFlushed` are left, in `MigrateLegacyToDescriptor` and `bench/wallet_migration.cpp`

achow101 replied:
Removed

mzumsande reviewed (commented):
I don't completely understand what the second part of the PR (`BESTBLOCK_NOMERKLE` serialization) achieves with the current approach.
With  748063a9820cb740c9b68f80b9e8c97d2a21e83d, `chainStateFlushed` has been made a no-op and could be removed.
On the other hand, the expressed goal of having `m_last_block_processed` and `m_last_block_processed_height` always in sync with the block locator is no longer current, given that we only update every 144 blocks if there is nothing relevant for the wallet in these blocks.
In that case, why not remove chainstateFlushed after 748063a9820cb740c9b68f80b9e8c97d2a21e83d (plus update the locator on unload)  and call it a day, avoiding any possible compatibility issues and the complications of a walletdb format change?

achow101 committed:
test, wallet: Remove concurrent writes test

Since CWallet::chainStateFlushed is now no-op, this test no longer tests
the concurrent writes scenario. There are no other cases where multiple
DatabaseBatches are open at the same time.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1520,18 +1542,25 @@ void CWallet::blockConnected(ChainstateRole role, const interfaces::BlockInfo& b
+
+    // Update on disk if this block resulted in us updating a tx, or periodically every 144 blocks (~1 day)
+    if (wallet_updated || block.height % 144 == 0) {
+        SetLastBlockProcessed(block.height, block.hash);
```
Done

achow101 commented:
I've removed the `BestBlock` struct refactor and walletdb changes so that this only changes how `m_last_block_processed` is updated. Those changes will be put into a separate PR at a later time.

I also noticed that the concurrent writes test in wallet_descriptor.py no longer tests the issue since it relied on `chainStateFlushed()` writing to the wallet. I'm pretty sure that there are no other cases of concurrent writes, so I think it is fine to remove that test now.

ryanofsky reviewed (approved):
Code review ACK 4882e71403111f11651f97a6b1cf7d861e619d8f. This change seems ok now that it is not writing to disk on every block, and there are some nice cleanups here.

I think PR description should be rewritten to say what the benefits of this change are now that statements like "This PR changes the those last block fields to actually be in sync with the record stored on disk." are no longer true.

I guess my thoughts are:

- I like all the small cleanups here making operations more explicit.
- I like that `blockConnected` notification now calls `WriteBestBlock` to update the wallet locator whenever the block contains any relevant transactions. That seems like a clear win.
- I think it's probably good that `blockDisconnected` notification calls `WriteBestBlock` every block. Might be a little cleaner to only do it when the wallet is actually updated like `blockConnected`, but doubt that matters.
- Calling `WriteBestBlock` every 144 blocks seems like it might write too frequently during initial sync and not write frequently enough after that. Seems like it might be smarter to update the best block based on clock time like the rescan implementation, which seems to write it every 60 seconds.
- I don't like not updating the best block when the `chainStateFlushed` notification is received. It seems nice to be able to tell a daemon to flush all its state to disk at once, and have some high level control over flushing for efficiency so each component is not just doing its own thing. But this is a theoretical objection because I don't think `chainStateFlushed` was a practically useful notification before, even though it could have been.

Overall LGTM.

achow101 commented on src/wallet/load.cpp:
```diff
@@ -173,7 +173,7 @@ void FlushWallets(WalletContext& context)
-void StopWallets(WalletContext& context)
+void CloseWallets(WalletContext& context)
 {
     for (const std::shared_ptr<CWallet>& pwallet : GetWallets(context)) {
```
Dropped it as suggested.

I believe `CloseWallets` did actually have an effect in that it is the step that deletes the `/databases` directory for legacy wallets. However, this step is skipped by `unloadwallet`, and afaict, is mostly unnecessary. So I've gone ahead and removed it. Since BDB is going to be removed, I see no reason to try to preserve this behavior especially when the two methods of unloading differ.

For sqlite wallets, this makes no difference.

ryanofsky reviewed (approved):
Code review ACK 480186d7093117932bb1f672932813a8936b6c42. Only change since last review is dropping extra wallet shutdown step, which seems like a nice simplification.

PR description has also been updated and looks good. I did find last two paragraphs a little confusing because when it mentions "last block" fields and records it sounds might be referring to `m_last_block_processed` and `m_last_block_processed_height` some places where it's actually referring to the best block locator. Would maybe suggest:

---

This PR changes those last block fields to be updated whenever the wallet makes a change to the db for new transaction state found in new blocks. Whenever a block is received that contains a transaction relevant to the wallet, the last block locator will now be written to disk. Furthermore, every block disconnection will now write an updated locator.

To ensure that the locator is relatively recent and loading rescans are fairly quick in the event of unplanned shutdown, it is also now written every 144 blocks (~1 day). Additionally it is now written when the wallet is unloaded so that it is accurate when the wallet is loaded again.

---

Could also add "The wallet locator is no longer written when a `chainstateFlushed` notification is received from the node, since it should already be mostly up to date"

achow101 commented:
Updated description mostly as suggested.

mzumsande reviewed (commented):
Looks good, except for an issue with commit 00fe040eca4b77d05ec5652b4d494b0ded183653

achow101 commented on src/wallet/load.h:
```diff
@@ -31,10 +31,10 @@ void StartWallets(WalletContext& context);
-//! Stop all wallets. Wallets will be flushed first.
-void StopWallets(WalletContext& context);
+//! Close all wallet databases.
+void CloseWallets(WalletContext& context);
```
Just forgot to remove it, done now.

The db is closed when a `CWallet` is destroyed.

mzumsande commented:
Looks like `WalletMigration` bench fails on this branch (which is now run in the CI).

achow101 committed:
wallet, bench: Write a bestblock record in WalletMigration

Migrating a wallet requires having a bestblock record. This is always
the case in normal operation as such a record is always written on
wallet loading if it didn't already exist. However, within the unit
tests and benchmarks, this is not guaranteed. Since migration requires
the record, WalletMigration needs to also add this record before the
benchmark.

achow101 commented:
Fixed CI

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3120,7 +3120,8 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         if (chain) {
-            walletInstance->chainStateFlushed(ChainstateRole::NORMAL, chain->getTipLocator());
+            WalletBatch batch(walletInstance->GetDatabase());
+            batch.WriteBestBlock(chain->getTipLocator());
```
In commit "wallet: Replace chainStateFlushed in loading with WriteBestBlock" (8b180358fb4c0585f2d9b8c4239d99a98c1fd663)

Not important, but if a goal of this PR is to try to keep m_last_block_processed and bestblock closer in sync, seems like it would be good to call SetLastBlockProcessed here instead of WriteBestBlock.

Same suggestion also applies to AttachChain below, but that WriteBestBlock to SetLastBlockProcessed replacement is already made in a later commit 08668272030bfcffc78efd77b642d6e50563d5dc. I think i'd suggest squashing that commit into this one to reduce internal churn in the PR.

achow101 replied:
Done and squashed the commits.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -164,6 +164,7 @@ bool RemoveWallet(WalletContext& context, const std::shared_ptr<CWallet>& wallet
 
     interfaces::Chain& chain = wallet->chain();
     std::string name = wallet->GetName();
+    wallet->WriteBestBlock();
```
In commit "wallet: Write best block record on unload" (eceb848494ea24d3686a9ff9a2d25b657cce9d6e)

Current change should be ok because it's ok if locator is a little behind, but this doesn't seem like most ideal place to write the best block because notifications could still be incoming at the point. A better place would seem like the FlushAndDeleteWallet function before the [Flush](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/wallet/wallet.cpp#L239) call. Even better might be to move both the WriteBestBlock call from here and the Flush call from FlushAndDeleteWallet into the CWallet destructor.

achow101 replied:
https://github.com/bitcoin/bitcoin/actions/runs/14892979963/job/41829335187?pr=30221 seems like the same possible issue that you've described where a notification comes in during the unload.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30221#discussion_r2073849511

Just to be clear I was suggesting moving both WriteBestBlock call and the [Flush](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/wallet/wallet.cpp#L239) call into the destructor, because if you only move the WriteBestBlock call, it will happen after flushing, which seems incorrectly logically, though maybe it is ok if sqlite Flush is a no-op. I didn't look into the CI failures though, just mentioning this in case it's related.

achow101 replied:
`Flush()` is deleted post #28710 as sqlite's `Flush()` is a no-op.

ryanofsky replied:
> `Flush()` is deleted post #28710 as sqlite's `Flush()` is a no-op.

Thanks, makes sense. I had been looking to see if the Flush call was moved in the diff, but of course it's not because it doesn't exist anymore.

achow101 replied:
I could not figure out the CI failure, so I've reverted this change.

ryanofsky replied:
In commit "wallet: Write best block record on unload" (8ecc7653556a1977eae5a902d51a2ececde94913)

Just to be sure, current state is that if WriteBestBlock is called here, rather than being called in the destructor or being called in FlushAndDeleteWallet, the best block locator could be a little out of date and this is maybe not ideal but fine?

re: https://github.com/bitcoin/bitcoin/pull/30221#issuecomment-2864638615

> It can't be static because the `FUZZ_TARGET` below needs to access it. However, the wallet could be created in there instead of in the setup, as the other wallet fuzz tests do.

Yes, suggestion would be to follow the fuzz test setup pattern and create a static variable and global pointer. But no need to change this if debugging the CI failure is hard and only consequence is the best block could be a little behind.

achow101 replied:
> the best block locator could be a little out of date and this is maybe not ideal but fine?

I'm not sure that actually could happen, but I think that's the worse that can happen, and I think that's fine. It should only be a couple of blocks which would be fast to rescan.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -791,7 +790,6 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     /** should probably be renamed to IsRelevantToMe */
     bool IsFromMe(const CTransaction& tx) const;
     CAmount GetDebit(const CTransaction& tx, const isminefilter& filter) const;
-    void chainStateFlushed(ChainstateRole role, const CBlockLocator& loc) override;
```
In commit "wallet: Remove chainStateFlushed" (2d72a8894cfb609ee616b098e165c722183ba325)

Would it be easily possible to squash this commit and the earlier commit "wallet: Remove WriteBestBlock from chainStateFlushed" (35b636469983fdcd5f8e62448f4733effe596e41) together?

I feel like that wouild make the PR easier to understand because the earlier commit doesn't provide any rationale while the commit message here is a little misleading because the commit at this point is only removing dead code.

Another alternative to squashing the two commits could be move to the rationale in this commit message to the earlier commit to describe reason for the earlier change.

achow101 replied:
Squashed.

ryanofsky reviewed (approved):
Code review ACK de7b5eda7ef04e875850fe594de2a89992b7bd46. No significant changes since the last review other than a rebase and a CI fix.

Left some suggestions that I think could make the PR a little easier to understand and review, but they are not important so feel free to ignore.

achow101 commented on test/functional/wallet_reorgsrestore.py:
```diff
@@ -122,11 +122,11 @@ def test_reorg_handling_during_unclean_shutdown(self):
-        # Due to an existing bug, the wallet incorrectly keeps the transaction in an abandoned state, even though that's
-        # no longer the case (after the unclean shutdown, the node's chain returned to the pre-invalidation tip).
-        # This issue blocks any future spending and results in an incorrect balance display.
+        # After disconnecting the block, the wallet should recorded the new best block.
```
Done

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30221#issuecomment-2861121747

> 🚧 At least one of the CI tasks failed. Task `fuzzer,address,undefined,integer, no depends`: https://github.com/bitcoin/bitcoin/runs/41829354924 LLM reason (✨ experimental): The CI failure is caused by a dynamic-type-mismatch detected by UndefinedBehaviorSanitizer during fuzz testing.

This seems unrelated to the other CI failures, but it looks like failure is caused by a bug in the fuzz test:

https://github.com/bitcoin/bitcoin/blob/7eed9f17aec3e3ba93ec180a072b1c2f08dcbb39/src/wallet/test/fuzz/fees.cpp#L17-L25

Where the CWallet object is destroyed after the TestingSetup object so the interfaces::Chain object is already destroyed and this error happens.

Simplest fix would probably be to make the CWallet object a static variable in the function like the TestingSetup variable is, instead of a global variable.

achow101 commented:
> Simplest fix would probably be to make the CWallet object a static variable in the function like the TestingSetup variable is, instead of a global variable.

It can't be static because the `FUZZ_TARGET` below needs to access it. However, the wallet could be created in there instead of in the setup, as the other wallet fuzz tests do.

But, with moving `WriteBestBlock` back to `RemoveWallet`, this is not necessary either.

mzumsande reviewed (commented):
Code Review ACK [e2e9c30](https://github.com/bitcoin/bitcoin/commit/e2e9c30a4d04bd8173c3386c5b97dbac227e810e)
(nits are not important)

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3225,13 +3229,21 @@ bool CWallet::AttachChain(const std::shared_ptr<CWallet>& walletInstance, interf
             }
+            walletInstance->m_attaching_chain = false;
+            // Set and update the best block record
+            // Although ScanForWalletTransactions will have stopped at the best block that was set prior to the rescan,
+            // we still need to make sure that the best block on disk is set correctly as rescanning may overwrite it.
```
In commit "wallet: Replace chainStateFlushed in loading with SetLastBlockProcessed" (357a65fd426f2122f390803d572fd89d4b0177c4)

I think everything in the comment is accurate but also vague, and it raises questions about when the last scanned block would not match the last/best blocks. I think it might be better to be more concrete like:

// Set last block scanned as the last block processed, because it may be different in the case of a reorg. Also save the best block locator because rescanning only updates it intermittently.

achow101 replied:
Done as suggested.

ryanofsky reviewed (approved):
Code review ACK e2e9c30a4d04bd8173c3386c5b97dbac227e810e just moving WriteBestBlock call from destructor back to RemoveWallet() since last review.

I left a bunch of suggestions below and repeated some of Martin's but they are all nitpicking at this point so feel free to ignore

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1573,6 +1586,9 @@ void CWallet::blockDisconnected(const interfaces::BlockInfo& block)
     }
+
+    // Update the best block
+    SetLastBlockProcessed(block.height - 1, *Assert(block.prev_hash));
```
I think I will leave this as is.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4471,4 +4472,17 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
     return std::nullopt;
 }
+
+void CWallet::WriteBestBlock() const
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1473,18 +1495,25 @@ void CWallet::blockConnected(ChainstateRole role, const interfaces::BlockInfo& b
+
+    // Update on disk if this block resulted in us updating a tx, or periodically every 144 blocks (~1 day)
+    if (wallet_updated || block.height % 144 == 0) {
+        SetLastBlockProcessed(block.height, block.hash);
```
Changed to `WriteBestBlock()`.

>  I also wonder if we can consolidate the seperate updates in blockConnected, deleting SetLastBlockProcessedInMem at the top, and moving it to the bottom here like:

It's at the top because there's some stuff in `AddToWallet` which need the last block processed in memory to be updated to the current block.

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -4453,4 +4477,17 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
     return std::nullopt;
 }
+
+void CWallet::WriteBestBlock() const
+{
+    AssertLockHeld(cs_wallet);
+
+    if (!m_last_block_processed.IsNull()) {
+        CBlockLocator loc;
+        chain().findBlock(m_last_block_processed, FoundBlock().locator(loc));
+
+        WalletBatch batch(GetDatabase());
+        batch.WriteBestBlock(loc);
+    }
+}
```
Now that we have this `WriteBestBlock` function, something similar happens within `BackupWallet` as well: https://github.com/bitcoin/bitcoin/blob/8309a9747a8df96517970841b3648937d05939a3/src/wallet/wallet.cpp#L3267-L3274

Do you think we can leverage `WriteBestBlock` over there? I mention because the pattern is quite similar and I would prefer to have the same function being called.

achow101 replied:
Indeed, done.

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -3220,13 +3224,21 @@ bool CWallet::AttachChain(const std::shared_ptr<CWallet>& walletInstance, interf
+            // Set and update the best block record
+            // Set last block scanned as the last block processed as it may be different in case the case of a reorg.
+            // Also save the best block locator because rescanning only updates it intermittently.
+            walletInstance->SetLastBlockProcessed(*scan_res.last_scanned_height, scan_res.last_scanned_block);
```
This comment is specific to the concept of scanning for wallet transactions and then updating the last block processed both in memory and disk:

> Additionally, after rescanning on wallet loading, instead of writing
the locator for the current chain tip, write the locator for the last
block that the rescan had scanned.

There are couple other usages of `ScanForWalletTransactions` in `rescanblockchain` RPC and conditionally in `importdescriptors`. I guess we don't want to update the block locator in disk in these cases because we also sync the block record to disk every 144 blocks now & doing it in these 2 RPCs doesn't seem necessary?

achow101 replied:
Those usages of `ScanForWalletTransactions` specifically set the option to not write the last block scanned to disk. This is because they are expected to be used during normal operation when the wallet will still be handling incoming blocks and transactions, so having that record possibly be out sync is actually quite bad.

mzumsande replied:
Also just mentioning that `rescanblockchain` can be used with ranges (`start_height` / `stop_height`). If  `stop_height` is set to some past block, setting the best block back to that old block would be incorrect and lead to wrong balances etc.

rkrux reviewed (commented):
This partial review is for the previous diff at e2e9c30a4d04bd8173c3386c5b97dbac227e810e. I started looking into it today and need to spend some more time in internalising the consequences of this diff but at the moment I'm inclining towards a concept ack.

Left few questions below.

rkrux commented on test/functional/wallet_reorgsrestore.py:
```diff
@@ -119,11 +119,11 @@ def test_reorg_handling_during_unclean_shutdown(self):
-        # Due to an existing bug, the wallet incorrectly keeps the transaction in an abandoned state, even though that's
-        # no longer the case (after the unclean shutdown, the node's chain returned to the pre-invalidation tip).
-        # This issue blocks any future spending and results in an incorrect balance display.
+        # After disconnecting the block, the wallet should record the new best block.
```
> the wallet should record the new best block.

I assume there is no way for us to verify this before the crash and after tip invalidation. The wallet info returns value based on the one that's in memory: `m_last_block_processed `

> wallet.getwalletinfo()['lastprocessedblock']['hash']

achow101 replied:
Not sure what you would be verified? After invalidateblock and before the crash, we do verify that the coinbase transaction is marked abandoned. This check is to verify that starting after the crash, the wallet has the correct before-invalidate state.

rkrux replied:
> Not sure what would be verified? After invalidateblock and before the crash, we do verify that the coinbase transaction is marked abandoned. 

I wanted to check that the best block locator is indeed persisted to disk after invalidation but the `getwalletinfo` RPC reads from memory, which can't be used for verifying this PR's change. The coinbase transaction verification was done before this PR as well, so that also not verifies.

Previously, I had in mind some functionality that reads from disk while returning the response in RPC but that might be going a bit too far just for tests. The other easier way I found out is to verify using the debug log from `AttachChain` that confirms the rescan indeed does happen - which tells me that the best block locator was persisted to disk after block invalidation and before node crash.
https://github.com/bitcoin/bitcoin/blob/3f83c744ac28b700090e15b5dda2260724a56f49/src/wallet/wallet.cpp#L3192


**Functional test diff that worked in my system:**
```diff
@@ -104,6 +104,7 @@ class ReorgsRestoreTest(BitcoinTestFramework):
 
         # Disconnect tip and sync wallet state
         tip = wallet.getbestblockhash()
+        tip_height = wallet.getblockstats(tip)["height"]
         wallet.invalidateblock(tip)
         wallet.syncwithvalidationinterfacequeue()
 
@@ -116,7 +117,8 @@ class ReorgsRestoreTest(BitcoinTestFramework):
         node.kill_process()
 
         # Restart the node and confirm that it has not persisted the last chain state changes to disk
-        self.start_node(0)
+        with self.nodes[0].assert_debug_log(expected_msgs=[f"Rescanning last 1 blocks (from block {tip_height - 1})...\n"]):
+            self.start_node(0)
         assert_equal(node.getbestblockhash(), tip)
 
         # After disconnecting the block, the wallet should record the new best block.
```

achow101 replied:
Ah, I see. I think that's unnecessary and the current test is sufficient to verify the behavior.

rkrux reviewed (commented):
ACK 30a94b1ab9ae850d55cb9eb606a06890437bc75e

I like how this PR brings the best/last block processed updates both in memory & disk in unison, making it easier to reason about the wallet code ultimately. The detailed commit messages aids review.

I have verified that the location of such changes makes it possible for the following wallet RPCs to correctly handle best block related memory & disk updates: 
```
createwallet
loadwallet
restorewallet
unloadwallet
backupwallet
```

I feel that updating the locator when wallet transactions are found in the latest block & periodically every 144 blocks in `blockConnected` notifications, and also going back to the previous best block hash in the corresponding `blockConnected` notification makes it more intuitive & easier to reason about. There's a possibility that the best block locator will be updated frequently if the wallet has transactions in contiguous blocks but I don't think that's a big deal as it's relatively less likely to occur. 

Even though I had not read the `chainStateFlushed` code prior to this PR, I found consistent usage of that terminology within wallet instance slightly distracting & thus support its removal.

I prefer the `SetLastBlockProcessedInMem` naming as it clearly mentions the storage that's being updated. Also in favour of `SetLastBlockProcessed` function updating in memory & disk in one go.

I realised that there's an opportunity to put `m_last_block_processed` & `m_last_block_processed_height` together in a struct as they are more often than not used together, I see a related refactor has been alluded to in this comment: https://github.com/bitcoin/bitcoin/pull/30221#issuecomment-2798088018

mzumsande reviewed (commented):
Code Review ACK 30a94b1ab9ae850d55cb9eb606a06890437bc75e

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3220,13 +3224,21 @@ bool CWallet::AttachChain(const std::shared_ptr<CWallet>& walletInstance, interf
 
```
Will do if I need to retouch.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4453,4 +4470,17 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
+
+    if (!m_last_block_processed.IsNull()) {
+        CBlockLocator loc;
+        chain().findBlock(m_last_block_processed, FoundBlock().locator(loc));
```
In commit "wallet: Update best block record after block dis/connect" (7bacabb204b6c34f9545f0b37e2c66296ad2c0de)

IMO would be good to assert `!loc.IsNull()` as a sanity check to ensure that doesn't happen. (It shouldn't, but node or wallet code could change in the future and invalidate current assumptions.)

ryanofsky reviewed (approved):
Code review ACK 30a94b1ab9ae850d55cb9eb606a06890437bc75e. Only changes since last review are using WriteBestBlock method more places and updating comments.

It looks like there are some small review comments above that could be followed up, but with 3 acks I think it would be good to merge this shortly and let smaller improvements can be made later. I plan to merge this soon after testing locally.

Dorex45 reviewed (commented):
**Thanks for working on syncing wallet state!**
 Concept ACK – wrapping block and height in a BestBlock struct makes sense.
Just A few thoughts:
• Could we add short comments explaining BestBlock members for clarity?
• In DeserializeBestBlock, should we warn/log if height is missing for backward compatibility?
• What about adding a test for a scenario involving a reorg, to ensure BestBlock updates in both blockConnected and blockDisconnected paths?
• Optional: Would a ToString() or debug helper for BestBlock aid future debugging?
Overall this looks solid ,thanks for the cleanup!
